src/GramatykiGrafoweAGH/project/task4.py

```diff
@@ -8,22 +8,31 @@
 def P7(G: Graph) -> None:
     for E2s in G.get_duplicates_with_label('E'):
         for E2L, E2R in combinations(E2s, 2):
+            E1L_, E2L_, E3L_ = None, None, None
+            E1R_, E2R_, E3R_ = None, None, None
+
             E13Ls = G.get_neighbors_with_label(E2L, 'E')
             for E1L, E3L in combinations(E13Ls, 2):
                 if is_node_between(E1L, E2L, E3L):
-                    for E1R in G.get_duplicates_of(E1L):
-                        for E3R in G.get_duplicates_of(E3L):
-                            if is_node_between(E1R, E2R, E3R):
-                                for I12L in G.get_common_neighbors_with_label(E1L, E2L, 'I'):
-                                    for I23L in G.get_common_neighbors_with_label(E2L, E3L, 'I'):
-                                        for iL in G.get_common_neighbors_with_label(I12L, I23L, 'i'):
-                                            for I12R in G.get_common_neighbors_with_label(E1R, E2R, 'I'):
-                                                for I23R in G.get_common_neighbors_with_label(E2R, E3R, 'I'):
-                                                    for iR in G.get_common_neighbors_with_label(I12R, I23R, 'i'):
-                                                        for E in G.get_common_neighbors_with_label(iL, iR, 'E'):
-                                                            G.merge_two_nodes(E1L, E1R)
-                                                            G.merge_two_nodes(E2L, E2R)
-                                                            G.merge_two_nodes(E3L, E3R)
-                                                            return
+                    E1L_, E2L_, E3L_ = E1L, E2L, E3L
+
+            if E1L_:
```

Comment:
```suggestion
            if E1L_ is not None:
```

Reply:
BTW, why these changes were necessary?

Reply:
They were not. As you said "Would be really nice if you find a way to refactor the nested for-loops so the code looks more elegant." an it is an attempt to refactor the code somewhat.